include/pybind11/numpy.h

```diff
@@ -163,11 +216,121 @@ template <typename T> constexpr const int npy_format_descriptor<
+            return object(ptr, true); \
+        pybind11_fail("Unsupported buffer format!"); \
+    } \
     static PYBIND11_DESCR name() { return _(Name); } }
-DECL_FMT(float, NPY_FLOAT_, "float32"); DECL_FMT(double, NPY_DOUBLE_, "float64"); DECL_FMT(bool, NPY_BOOL_, "bool");
-DECL_FMT(std::complex<float>, NPY_CFLOAT_, "complex64"); DECL_FMT(std::complex<double>, NPY_CDOUBLE_, "complex128");
+DECL_FMT(float, NPY_FLOAT_, "float32");
+DECL_FMT(double, NPY_DOUBLE_, "float64");
+DECL_FMT(bool, NPY_BOOL_, "bool");
+DECL_FMT(std::complex<float>, NPY_CFLOAT_, "complex64");
+DECL_FMT(std::complex<double>, NPY_CDOUBLE_, "complex128");
 #undef DECL_FMT
 
+struct field_descriptor {
+    const char *name;
+    int offset;
+    object descr;
+};
+
+template <typename T>
+struct npy_format_descriptor<T, typename std::enable_if<is_pod_struct<T>::value>::type> {
+    static PYBIND11_DESCR name() { return _("user-defined"); }
+
+    static object dtype() {
+        if (!dtype_())
+            pybind11_fail("NumPy: unsupported buffer format!");
+        return object(dtype_(), true);
+    }
+
+    static const char* format() {
+        if (!dtype_())
+            pybind11_fail("NumPy: unsupported buffer format!");
+        return format_();
+    }
+
+    static void register_dtype(std::initializer_list<field_descriptor> fields) {
+        array::API& api = array::lookup_api();
+        auto args = dict();
+        list names { }, offsets { }, formats { };
+        for (auto field : fields) {
+            if (!field.descr)
+                pybind11_fail("NumPy: unsupported field dtype");
+            names.append(str(field.name));
+            offsets.append(int_(field.offset));
+            formats.append(field.descr);
+        }
+        args["names"] = names;
+        args["offsets"] = offsets;
+        args["formats"] = formats;
+        // This is essentially the same as calling np.dtype() constructor in Python and passing
+        // it a dict of the form {'names': ..., 'formats': ..., 'offsets': ...}.
+        if (!api.PyArray_DescrConverter_(args.release().ptr(), &dtype_()) || !dtype_())
+            pybind11_fail("NumPy: failed to create structured dtype");
+        // Let NumPy figure the buffer format string for us: memoryview(np.empty(0, dtype)).format
+        auto np = module::import("numpy");
+        auto empty = (object) np.attr("empty");
+        if (auto arr = (object) empty(int_(0), dtype())) {
+            if (auto view = PyMemoryView_FromObject(arr.ptr())) {
+                if (auto info = PyMemoryView_GET_BUFFER(view)) {
+                    std::strncpy(format_(), info->format, 4096);
+                    return;
+                }
+            }
+        }
+        pybind11_fail("NumPy: failed to extract buffer format");
+    }
+
+private:
+    static inline PyObject*& dtype_() { static PyObject *ptr = nullptr; return ptr; }
+    static inline char* format_() { static char s[4096]; return s; }
```

Comment:
Well, it depends... In C++ you will still work with struct A, but on the Python side you would either see structs that will either see the `a` part or the `b` part. I can't see a particular use case for this though, and since we essentially glue the dtype to the C++ type, maybe it makes sense to ensure it's called at most once for each C++ type?..

Ha, btw I believe this may have pointed me at a potential bug, if you do a `PYBIND11_DTYPE(A, a)`, I think it will set itemsize incorrectly (because I forgot to pass `sizeof(A)`) to dtype initializer. I'll fix it and add some tests.